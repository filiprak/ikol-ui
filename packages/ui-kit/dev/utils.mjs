import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
