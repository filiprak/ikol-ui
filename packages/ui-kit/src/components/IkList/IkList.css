.ik-list-item {
    align-items: center;
    display: grid;
    grid-template-areas: "prepend content append";
    grid-template-columns: max-content auto max-content;
    outline: none;
    max-width: 100%;
    padding: 8px 12px;
}

a.ik-list-item {
    text-decoration: none;
    color: inherit;
}

.ik-list-item--inline {
    display: inline-grid;
    display: -ms-inline-grid;
    display: -moz-inline-grid;
}

.ik-list-item--no-select {
    user-select: none !important;
}

.ik-list-item--no-padding {
    padding: 0;
}

.ik-list-item--disabled {
    pointer-events: none !important;
    user-select: none !important;
    cursor: not-allowed !important;
    opacity: 0.65 !important;
}

.ik-list-item--align-start > .ik-list-item__prepend,
.ik-list-item--align-start > .ik-list-item__content,
.ik-list-item--align-start > .ik-list-item__append {
    align-self: start;
}

.ik-list-item--clickable {
    cursor: pointer !important;
}

.ik-list-item--border-bottom {
    border-bottom: 1px solid var(--border-color-1);
}

.ik-list-item--active,
.ik-list-item--clickable:not(.ik-list-item--no-hover):hover {
    background-color: var(--on-background-color-5);
}

.ik-list-item__prepend {
    display: inline-grid;
    align-items: center;
    align-self: center;
    grid-area: prepend;
    padding-right: 10px;
}

.ik-list-item__content {
    align-self: center;
    grid-area: content;
    overflow: hidden;
}

.ik-list-item__append {
    display: inline-grid;
    align-items: center;
    align-self: center;
    grid-area: append;
    padding-left: 10px;
}

.ik-list-item__secondary {
    font-size: 0.85em;
    color: var(--on-background-color-3);
    font-weight: 400;
    overflow: hidden;
    text-overflow: ellipsis;
}

.ik-list-item:not(.ik-list-item--wrap-text) > .ik-list-item__content > .ik-list-item__title,
.ik-list-item:not(.ik-list-item--wrap-text) > .ik-list-item__content > .ik-list-item__secondary {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.ik-list-item:not(.ik-list-item--wrap-text).ik-list-item--lines > .ik-list-item__content > .ik-list-item__secondary {
    white-space: normal;
    -webkit-box-orient: vertical;
    display: -webkit-box;
    overflow: hidden;
    text-overflow: ellipsis;
}

.ik-list-item:not(.ik-list-item--wrap-text).ik-list-item--lines-1 > .ik-list-item__content > .ik-list-item__secondary {
    -webkit-line-clamp: 1;
}

.ik-list-item:not(.ik-list-item--wrap-text).ik-list-item--lines-2 > .ik-list-item__content > .ik-list-item__secondary {
    -webkit-line-clamp: 2;
}

.ik-list-item:not(.ik-list-item--wrap-text).ik-list-item--lines-3 > .ik-list-item__content > .ik-list-item__secondary {
    -webkit-line-clamp: 3;
}

.ik-list-item:not(.ik-list-item--wrap-text).ik-list-item--lines-4 > .ik-list-item__content > .ik-list-item__secondary {
    -webkit-line-clamp: 4;
}

.ik-list-item:not(.ik-list-item--wrap-text).ik-list-item--lines-5 > .ik-list-item__content > .ik-list-item__secondary {
    -webkit-line-clamp: 5;
}