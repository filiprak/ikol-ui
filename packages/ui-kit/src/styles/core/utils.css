/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
