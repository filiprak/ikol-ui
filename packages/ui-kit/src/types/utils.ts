import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
