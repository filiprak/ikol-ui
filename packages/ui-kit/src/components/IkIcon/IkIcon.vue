<script setup lang="ts">
import '@ui/styles';
import './IkIcon.css';
import { type PropType, h, useAttrs } from 'vue';
import { toCssUnits, getIconClasses } from '@ui/utils/helpers';
import { useRender } from '@ui/composables/render';
import type { IkUIDesignColor } from '@ui/types/utils';

const props = defineProps({
    tag: {
        type: String,
        default: 'i',
    },
    icon: {
        type: String,
    },
    size: {
        type: [Number, String],
    },
    size_px: {
        type: [Number, String],
    },
    disabled: {
        type: Boolean,
        default: false,
    },
    circle: {
        type: Boolean,
        default: false,
    },
    design: {
        type: String as PropType<`${IkUIDesignColor}`>,
    },
    variant: {
        type: String,
    },
});
const attrs = useAttrs();

useRender(() => {
    return h(props.tag, {
        class: [
            {
                'ik-icon': true,
                ['ik-icon--' + props.design]: !!props.design,
                ['ik-icon--' + props.variant]: !!props.variant,
                'ik-icon--circle': props.circle,
                'ik-icon--clickable': !props.disabled && !!attrs.onClick,
                'ik-icon--disabled': props.disabled,
                ['fa-' + props.size + 'x']: !!props.size,
            },
            getIconClasses(props.icon || ''),
        ],
        style: {
            width: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            height: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            fontSize: props.size_px ? toCssUnits(props.circle ? (Number(props.size_px) / 2) : props.size_px, 'px') : null,
        },
    });
});
</script>
