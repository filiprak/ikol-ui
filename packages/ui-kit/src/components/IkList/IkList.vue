<template>
    <div class="ik-list">
        <slot></slot>
    </div>
</template>
<script setup lang="ts">
import '@ui/styles';
import './IkList.css';
</script>
