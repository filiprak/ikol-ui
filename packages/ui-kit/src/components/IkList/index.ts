import IkList from './IkList.vue';
import IkListItem from './IkListItem.vue';
export interface IkList extends InstanceType<typeof IkList> { };
export interface IkListItem extends InstanceType<typeof IkListItem> { };
export { IkListItem, IkList };
