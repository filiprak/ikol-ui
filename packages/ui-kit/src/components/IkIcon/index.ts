import IkIcon from './IkIcon.vue';
export interface IkIcon extends InstanceType<typeof IkIcon> { };
export { IkIcon };
