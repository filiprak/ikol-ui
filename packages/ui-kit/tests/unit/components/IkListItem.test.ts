import { renderShallow } from '@ui/tests/tools/utils';
import { IkListItem } from '@ui/components/IkList';
import { IkIcon } from '@ui/components/IkIcon';

describe('IkListItem', () => {
    const getInstance = (options = {}) => renderShallow(IkListItem, options);

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toStrictEqual([
            'ik-list-item',
        ]);
        expect(wrapper.element.tagName).toBe('DIV');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: '<div class="test">List item content slot.</div>',
            },
        });
        expect(wrapper.text()).toContain('List item content slot.');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with secondary slot', () => {
        const wrapper = getInstance({
            slots: {
                secondary: 'This is secondary content.',
            },
        });
        expect(wrapper.text()).toContain('This is secondary content.');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders inline', () => {
        const wrapper = getInstance({
            props: {
                inline: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--inline');
    });

    it('renders active', () => {
        const wrapper = getInstance({
            props: {
                active: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--active');
    });

    it('renders with bottom border', () => {
        const wrapper = getInstance({
            props: {
                border_bottom: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--border-bottom');
    });

    it('renders clickable', () => {
        const clickFn = jest.fn();
        const wrapper = getInstance({
            props: {
                onClick: clickFn,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--clickable');
    });

    it('renders text wrapped', () => {
        const wrapper = getInstance({
            props: {
                wrap_text: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--wrap-text');
    });

    it('renders with icon prop', () => {
        const wrapper = getInstance({
            props: {
                icon: 'user:light',
            },
        });

        expect(wrapper.findAllComponents(IkIcon).at(0)?.attributes('icon')).toBe('user:light');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with disabled prop', () => {
        const wrapper = getInstance({
            props: {
                disabled: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-list-item--disabled');
    });

    it('renders with append slot', () => {
        const wrapper = getInstance({
            slots: {
                append: '<span>Append slot...</span>',
            },
        });

        expect(wrapper.text()).toContain('Append slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with prepend slot', () => {
        const wrapper = getInstance({
            slots: {
                prepend: '<span>Prepend slot...</span>',
            },
        });

        expect(wrapper.text()).toContain('Prepend slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with link option', async () => {
        const wrapper = getInstance({
            props: {
                link: '/test?unit=true',
            },
        });

        expect(wrapper.element.tagName).toBe('A');
        expect(wrapper.find('a').attributes('href')).toBe('/test?unit=true');
        expect(wrapper.classes()).toContain('ik-list-item--clickable');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders not text-selectable', () => {
        const wrapper = getInstance({
            props: {
                no_select: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--no-select');
    });

    it('renders without padding', () => {
        const wrapper = getInstance({
            props: {
                no_padding: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--no-padding');
    });

    it('renders without hover effect', () => {
        const wrapper = getInstance({
            props: {
                no_hover: true,
            },
        });
        expect(wrapper.classes()).toContain('ik-list-item--no-hover');
    });

    it('renders aligned to the top', () => {
        const wrapper = getInstance({
            props: {
                align_start: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-list-item--align-start');
    });

    it('renders text wrapped to n lines', () => {
        const wrapper = getInstance({
            props: {
                lines: 3,
            },
        });

        expect(wrapper.classes()).toContain('ik-list-item--lines');
        expect(wrapper.classes()).toContain('ik-list-item--lines-3');
    });
});
