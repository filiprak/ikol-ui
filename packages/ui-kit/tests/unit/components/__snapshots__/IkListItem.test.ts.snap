// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkListItem renders with append slot 1`] = `
<div
  class="ik-list-item"
>
  <!--v-if-->
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
    </div>
    <!--v-if-->
  </div>
  <div
    class="ik-list-item__append"
  >
    
    <span>
      Append slot...
    </span>
    
  </div>
</div>
`;

exports[`IkListItem renders with correct default state 1`] = `
<div
  class="ik-list-item"
>
  <!--v-if-->
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
    </div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`IkListItem renders with default slot 1`] = `
<div
  class="ik-list-item"
>
  <!--v-if-->
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      <div
        class="test"
      >
        List item content slot.
      </div>
      
    </div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`IkListItem renders with icon prop 1`] = `
<div
  class="ik-list-item"
>
  <div
    class="ik-list-item__prepend"
  >
    
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="user:light"
      tag="i"
    />
    
  </div>
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
    </div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`IkListItem renders with link option 1`] = `
<a
  class="ik-list-item ik-list-item--clickable"
  href="/test?unit=true"
>
  <!--v-if-->
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
    </div>
    <!--v-if-->
  </div>
  <!--v-if-->
</a>
`;

exports[`IkListItem renders with prepend slot 1`] = `
<div
  class="ik-list-item"
>
  <div
    class="ik-list-item__prepend"
  >
    
    <span>
      Prepend slot...
    </span>
    
  </div>
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
    </div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`IkListItem renders with secondary slot 1`] = `
<div
  class="ik-list-item"
>
  <!--v-if-->
  <div
    class="ik-list-item__content"
  >
    <div
      class="ik-list-item__title"
    >
      
      
    </div>
    <div
      class="ik-list-item__secondary"
    >
      
      This is secondary content.
      
    </div>
  </div>
  <!--v-if-->
</div>
`;
